### ISSUE TITLE
Model summary export into CSV

### ISSUE THREAD

**Issue by TarasIakymchuk** [larq/larq#215]
### Feature motivation

Having model summary in CSV will allow to use these data in performance estimations or energy models, as well it's easier to include in publications with arbitrary text style. CSV is very easy to manipulate and is supported by all programming languages, CAS or spreadsheet editors.
### Feature description

Add a function that will export the model summary into CSV file with the name given by the user.

**Comment by Tombana** [comment]
I wrote a function `larq.save_summary(model, "filename.csv")` and currently it outputs the following. 

@TarasIakymchuk  This is probably not yet exactly what you want, so could you describe how it should be different from this?

```
Layer,Input prec. (bit),Outputs,# 1-bit x 1,# 2-bit x 1,# 32-bit x 1,Memory (kB),1-bit MACs,2-bit MACs,32-bit MACs
quant_conv2d,-,"(-1, 64, 64, 32)",288,0,32,0.16015625,0,0,1179648
max_pooling2d,-,"(-1, 32, 32, 32)",0,0,0,0,0,0,0
quant_depthwise_conv2d,2,"(-1, 11, 11, 32)",0,288,0,0.0703125,0,34848,0
quant_separable_conv2d,1,"(-1, 11, 11, 32)",1312,0,32,0.28515625,158752,0,0
flatten,-,"(-1, 3872)",0,0,0,0,0,0,0
dense,-,"(-1, 10)",0,0,38730,151.2890625,0,0,38720
Total,,,1600,288,38794,151.8046875,158752,34848,1218368

Total params,Trainable params,Non-trainable params,Model size,Float-32 Equivalent,Compression Ratio of Memory,Number of MACs,Ratio of MACs that are binarized,Ratio of MACs that are ternarized
40.7 k,1.95 k,38.7 k,0.15 MB,0.16 MB,0.9552627697753306,1.41 M,0.1124,0.0247
```

**Comment by TarasIakymchuk** [comment]
Not bad, not bad! I can import it into my Calc sheet and get the data machine readable. The only thing I will ask is to break the (-1,64,64,32) into separate cells and remove the useless -1. Does the memory count include both weights and activations or only the weights?

**Comment by TarasIakymchuk** [comment]
This is very close to what I want. Please split the (-1, 11, 11, 32) into separate cells and remove the useless -1.

**Comment by Tombana** [comment]
The reason I did not break the output sizes into separate cells, is because the dimensions change per layer. So the first layer has four numbers `(-1, 64, 64, 32)` and the last layer ony has two `(-1, 10)`. So then things would not be aligned anymore. A possible solution is to always have four cells and just leave them empty when there are less than four numbers. (Or three cells when removing the `-1`). Is that ok?

> Does the memory count include both weights and activations or only the weights?

I don't know to be honest. Currently I use exactly the same output that is generated by the existing `summary()` function.

**Comment by lgeiger** [comment]
I agree with @Tombana, I think breaking down the tuple into own cells would be very confusing since `larq` supports many different dimensions. I'd also prefer to keep the `-1` since it stands for the variable batch size the model supports.
If this makes it hard to parse, we could also save it as `json` file or pandas dataframe instead.

**Comment by TarasIakymchuk** [comment]
Last layer has the dimension of 1x1x10, PW convo is identical to the FC layer.

**Comment by Tombana** [comment]
> Last layer has the dimension of 1x1x10, PW convo is identical to the FC layer.

I'm not sure what you mean, but I think what @lgeiger is saying is that some models might even have more than four output dimensions, so always fixing 4 cells is not possible in general. Thats why its easier to have it as a string in its own cell. I can change the format within the string if you prefer.

**Comment by TarasIakymchuk** [comment]
More than 4 dims? Maybe 3D CNNs only. I think this case is very rare.

**Comment by lgeiger** [comment]
> More than 4 dims? Maybe 3D CNNs only. I think this case is very rare.

It might be rare for standard image recognition, but we support arbitrary kernel dimensions and custom layers, so `model.summary()` needs to handle this case as well.

**Comment by Tombana** [comment]
We can leave `summary()` as it is, but clip the result of `save_summary()` to output at most 4 dimensions ?  Its true that this would be rare, and we could for example throw an error in `save_summary()` when its more than 4 dimensions.

**Comment by TarasIakymchuk** [comment]
Redundancy for the universality.

**Comment by Tombana** [comment]
I fully agree with @timdebruin since the CSV output is more for internal use.

**Comment by TarasIakymchuk** [comment]
Let then agree on 3D activations and close this issue.

**Comment by Tombana** [comment]
Its implemented in [this branch](https://github.com/larq/larq/compare/csv_summary) on larq. Now it still has to be moved to `plumerai-utils`. It still depends on internals of `larq/models.py` though, because otherwise that code would be duplicated. Any thoughts on how to solve that? @lgeiger @timdebruin

**Comment by timdebruin** [comment]
> It still depends on internals of larq/models.py though, because otherwise that code would be duplicated. Any thoughts on how to solve that?
That should not be an issue as the energy calculations are also just an extension of things defined in `larq/models.py`.

**Comment by TarasIakymchuk** [comment]
I have one more request. How hard will be to split MACs into conv ops and non-conv ops? This is important for me. Or you can add the kernel size into the summary,  I can calculate conv MACs from it.
